044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,342 @@
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster. We intend to continue
+using `client.quota.callback.static.storage.check-interval` to configure the millisecond interval between polling the
+cluster state (default is 0 - disabled). The performance cost of this API was [discussed](https://lists.apache.org/thread/11zyqqnyg1wgf4jdo6pvn7hn51g3vf8r) 
+upstream as part of the KIP, which should be low cost.
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions through the admin api we need to construct an admin client.
+
+If using cluster sourced volumes we require the admin client bootstrap to be configured using `client.quota.callback.static.kafka.admin.bootstrap.servers`
+
+Additional admin client configuration can be passed using the form `client.quota.callback.static.kafka.admin.${configuration}`.
+For example: `client.quota.callback.static.kafka.admin.security.protocol=SSL`
+
+The strimzi-kafka-operator would be responsible for configuring the admin client to connect to the
+replication listener of the broker.
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+used bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume. When
+that limit is exceeded we effectively stop messages being produced.
+
+If the plugin is configured to use a cluster volume source, we also require one limit to be configured.
+
+Defining multiple limit types would be an invalid state.
+
+The limits we want are:
+1. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+2. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+For example, to configure a limit when availableBytes is below 1GB:
+- `client.quota.callback.static.storage.per.volume.limit.min.available.bytes=1000000000`
+
+Another example, to configure a limit when availableRatio is below 0.01 (1%):
+- `client.quota.callback.static.storage.per.volume.limit.min.available.ratio=0.01`
+
+Defining more than one limit is **invalid**, for example the below is disallowed:
+- `client.quota.callback.static.storage.per.volume.limit.min.available.ratio=0.05`
+- `client.quota.callback.static.storage.per.volume.limit.min.available.bytes=5000000000`
```

Comment:
How would the configuration look like if one day we add support for some _overrides_ for specific volume on specific brokers? Would this allow to add such feature in the future?

Reply:
Do you have an example of a use case? We can see that overrides would have been useful when the limits were based on used bytes. We are having a harder time imagining when you would want to do this with minimum available bytes.

Reply:
Not really. And I'm not suggesting to support it as part of this proposal. Just anticipating the future needs. It would be good to have an idea how it might fit into the config without any compatibility issues. I assume we could possibly do something similar to what Kafka does for listeners?

Reply:
Yeah I could imagine supporting overrides through the properties in future, maybe with an index to give the override rules a precedence/weighting. Like:

```
...volume.limit.1.broker=0
...volume.limit.1.logdirpath=/data/1
...volume.limit.1.min.available.ratio=0.02
```

or with some ugly csv like in this [authorizer plugin](https://github.com/bf2fc6cc711aee1a0c2a/kas-broker-plugins#customaclauthorizer)

Reply:
Ok, that sounds good I guess. Fingers crossed it will be never needed ;-)